src: tag: Add new feature 'kw tag'

BLBispo opened the pull request:
**Purpose**

Often a kernel developer has to perform some operations over commit or patch trailer lines. Adding tags such as 'Reviewed-by' or 'Acked-by' is a common task and sometimes it might be required to do this over multiple commits or patches. Also another common one is the 'Fixes' tag, that must follow a specific format.

This feature helps on these cases. It uses 'git rebase', 'git commit --amend' and 'git interpret-trailers' as its backend in order to make these tasks quicker to complete.

**Inspiration**

Both https://github.com/kworkflow/kworkflow/issues/1049 and https://github.com/kworkflow/kworkflow/issues/1051 represent similar issues, regarding trailer line manipulation of commits and patches. The idea is to deliver a new feature that can make such tasks easier and quicker to deal with.

This PR builds upon the work initiated in  https://github.com/kworkflow/kworkflow/pull/1121

**Some known limitations**

There is no way to "revert" any change made by this command through kw. For example, if the commit log changes, there is no command to undo the previous operation. User will have to manually use git reset --hard if changes are not what they expected.

This feature currently does not provide ways to configure conditional actions, which is possible via git config as shown by Melissa here: https://github.com/kworkflow/kworkflow/issues/1049

BLBispo commented:
Based on the comments from https://github.com/kworkflow/kworkflow/pull/1121, I made the following changes:
- Changed the feature name to "Tag";
- Moved the Signed-Off and Tested tags to the beginning;
- Fixed the problem with the "Fixes" tag for SHA 51fc8079c75ea89500a2953d02dcd304719f6929.

I am still learning how to program in bash, so I apologize if something is wrong. Below are some comments I’m not sure how to correct or explain, and I would appreciate your feedback:

When I moved Signed-Off and Tested to the beginning, I forgot to update the tests, but before opening the PR, I corrected the tests and rebased the commit. I didn’t encounter any issues with the unit tests locally. However, I noticed that when I opened the PR, some test errors appeared, but I’m not sure how to fix them. Could someone help me identify what might be wrong?

When I fixed the "Fixes" tag, I used the SHA 521292d5f87d2c8c2df2d4e9a150a694f05316b8, because I couldn’t find the SHA 51fc8079c75ea89500a2953d02dcd304719f6929.

I didn’t modify the validate email function; I just moved it from send_patch.sh to kwlib.sh, so I didn’t investigate this part deeply.

rodrigosiqueira reviewed (changes requested):
Hi @BLBispo ,

First of all, thanks a lot for your PR. This new feature looks really nice.

I noticed that your unit tests are failing, could you address this issue? Also, I added a few inline comments, could you also address those?

P.s.: After you fix the unit test issues I'll make a more detailed review.

BLBispo committed:
src: lib: kwlib: Move validate email functions to kwlib

These functions might be useful outside of send_patch. Moving them
to kwlib will help to avoid code duplication in other features that
also want to perform some sort of email validation.

Signed-off-by: Lincoln Yuji <lincolnyuji@hotmail.com>
Signed-off-by: Bruna Bispo <blbispo1@gmail.com>

BLBispo committed:
src: tag: Add new feature 'kw tag'

Often a kernel developer has to perform some operations over
commit or patch trailer lines. Adding tags, such as 'Reviewed-by',
is a common task and sometimes it might be required to do this over
multiple commits or patches. Also another usual one is the 'Fixes'
tag, that must follow a specific format.

This feature helps on these cases. It uses 'git rebase',
'git commit --amend' and 'git interpret-trailers' as its
backend in order to make these tasks quicker to complete.

Co-authored-by: Luiza Soezima <lbrsoezima@usp.br>
Co-authored-by: Sabrina Araujo <sabrinaaraujo@usp.br>
Signed-off-by: Lincoln Yuji <lincolnyuji@hotmail.com>
Signed-off-by: Bruna Bispo <blbispo1@gmail.com>

BLBispo commented on documentation/conf.py:
```diff
@@ -181,6 +181,7 @@
     ('man/features/kw-self-update', 'kw-self-update', 'kw self-update mechanism', ['David Tadokoro, Everaldo Junior'], 1),
     ('man/features/kw-patch-hub', 'kw-patch-hub', 'Terminal UI to interact with patches from lore.kernel.org', ['David Tadokoro, Rodrigo Siqueira'], 1),
     ('man/features/kw-bd', 'kw-bd', 'build and then deploy the kernel', ['Briza Mel de Sousa, Lorenzo Salvador'], 1),
+    ('man/features/kw-tag', 'kw-tag', 'write trailer lines', ['Bruna Bispo', 'Lincoln Yuji, Luiza Soezima, Sabrina Araujo'], 1),
```
Sorry, you are correct. I misunderstood and thought it would be 'kw-tag,' so the name ended up being just 'tag.' I have corrected it to 'kernel-tag.'

BLBispo commented on src/bash_autocomplete.sh:
```diff
@@ -84,6 +84,9 @@ function _kw_autocomplete()
 
+  kw_options['tag']='--add-signed-off-by --add-reviewed-by --add-acked-by --add-fixes
+                           --add-tested-by --add-co-developed-by --add-reported-by --verbose --help'
+
```
I have added this option in the file kw.fish. I am not sure if this addresses your comment, if not, could you please provide more information about what I need to do?

BLBispo commented on src/lib/kwlib.sh:
```diff
@@ -752,3 +755,52 @@ function get_git_repository_branches()
   done <<< "$output"
 }
+
+# This function validates the encryption. If the passed encryption is not valid
```
Okay, done.

BLBispo commented on src/tag.sh:
Okay, done.

BLBispo commented on src/tag.sh:
```diff
@@ -0,0 +1,643 @@
+  local sha="$1"
+
+  if [[ $(git cat-file -t "$sha" 2> /dev/null) != 'commit' ]]; then
+    return 1 # EPERM
```
Okay, done.

BLBispo commented on documentation/man/features/kw-tag.rst:
```diff
@@ -0,0 +1,172 @@
+
+Adding an **Acked-by** line starting from ``HEAD~2`` until ``HEAD``::
+
+  $ kw tag --acked-by='Some Name <example@mail.com>' HEAD~3
```
Okay, corrected to '--add-acked-by'. Does this problem occur only with '--add-acked-by'? Could you please provide me with more information about this branch and the merge commits?

BLBispo replied:
@rodrigosiqueira could you please help me with this information?

rodrigosiqueira replied:
Try this:

Clone Linus repository:
```
git clone https://github.com/torvalds/linux.git
```

Create a new branch:
```
git checkout -b TEST
```

Try something like:
```
kw kernel-tag --add-acked-by="Juca Pirama <juca@pirma.nothing.la>"  HEAD~3
```

I just tried it again, but this time, it created some merge conflicts.

BLBispo replied:
Okay, done.

BLBispo commented:
The unit tests failed, but I didn’t encounter any issues with them locally again.

![kernel_tag_test](https://github.com/user-attachments/assets/0cd8d095-4e86-4ddc-8d02-2dd5e18ae82d)

rodrigosiqueira commented:
Hi,
This PR has been open for a long time, and it is my fault. I'm really sorry for that. I'm not sure if you are still interested in working on this PR, but I'll close it for now since it's been here for a while. If you decide to bring this change back, I'll put the review and support for this PR as a high priority.

Again, I'm really sorry for my lack of review here.
